smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/IdempotencyTokenIgnoredValidator.java

```diff
@@ -0,0 +1,131 @@
+        }
+        List<ValidationEvent> events = new ArrayList<>();
+        NeighborProvider reverse = NeighborProviderIndex.of(model).getReverseProvider();
+        for (MemberShape memberShape : model.getMemberShapes()) {
+            Shape container = model.expectShape(memberShape.getContainer());
+            // Skip non-structures (invalid) and mixins (handled at mixed site).
+            if (!container.isStructureShape() || container.hasTrait(MixinTrait.class)) {
+                continue;
+            }
+
+            if (memberShape.hasTrait(IdempotencyTokenTrait.class)) {
+                Trait trait = memberShape.expectTrait(IdempotencyTokenTrait.class);
+                events.addAll(checkRelationships(container.asStructureShape().get(), memberShape, trait, reverse));
+            }
+        }
+        return events;
+    }
+
+    private List<ValidationEvent> checkRelationships(
+        StructureShape containerShape,
+        MemberShape memberShape,
+        Trait trait,
+        NeighborProvider reverse
+    ) {
+
+        // Store relationships so we can emit one event per ignored binding.
+        Map<RelationshipType, List<ShapeId>> ignoredRelationships = new TreeMap<>();
+        List<ValidationEvent> events = new ArrayList<>();
+        List<Relationship> relationships = reverse.getNeighbors(containerShape);
+        int checkedRelationshipCount = relationships.size();
+        for (Relationship relationship : relationships) {
+            // Skip members of the container.
+            if (relationship.getRelationshipType() == RelationshipType.MEMBER_CONTAINER) {
+                checkedRelationshipCount--;
+                continue;
+            }
+            if (relationship.getRelationshipType() != RelationshipType.INPUT) {
+                ignoredRelationships.merge(relationship.getRelationshipType(),
+                                           ListUtils.of(relationship.getShape().getId()), this::mergeShapeIdLists);
+            }
+        }
+
+        // If we detected invalid relationships, build the right event message.
+        if (!ignoredRelationships.isEmpty()) {
+            events.add(emit(memberShape, trait, checkedRelationshipCount, ignoredRelationships));
+        }
+
+        return events;
+    }
+
+    private List<ShapeId> mergeShapeIdLists(List<ShapeId> shapeIds1, List<ShapeId> shapeIds2) {
+        List<ShapeId> shapeIds = new ArrayList<>();
+        shapeIds.addAll(shapeIds1);
+        shapeIds.addAll(shapeIds2);
+        return shapeIds;
+    }
+
+    private ValidationEvent emit(
+        MemberShape memberShape,
+        Trait trait,
+        int checkedRelationshipCount,
+        Map<RelationshipType, List<ShapeId>> ignoredRelationships
+    ) {
+        String mixedIn = memberShape.getMixins().isEmpty() ? "" : " mixed in";
+        String message = "The `%s` trait applied to this%s member is ";
+        if (checkedRelationshipCount == ignoredRelationships.size()) {
```

Comment:
Why omit the context here? It looks like we emit the event on the member that has the trait, but don't say how the containing shape was referenced in a way that caused the trait to be ignored.

Reply:
I will remove.